napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import annotations
+
+import logging
+from concurrent.futures import Executor, Future, ThreadPoolExecutor
+from threading import RLock
+from typing import Dict, Iterable, Optional, Tuple
+
+from napari.components import Dims
+from napari.layers import Layer
+from napari.utils.events.event import EmitterGroup, Event
+
+logger = logging.getLogger("napari.components._layer_slicer")
+
+
+class _LayerSlicer:
+    """
+    High level class to control the creation of a slice (via a slice request),
+    submit it (synchronously or asynchronously) to a thread pool, and emit the
+    results when complete.
+
+    Attributes
+    ----------
+    events : napari.utils.events.event.EmitterGroup
+        vispy event emitter
+    _executor : concurrent.futures.ThreadPoolExecutor
+        manager for the slicing threading
+    _layers_to_task : dict
+        task storage for cancellation logic
+    lock_layers_to_task : threading.RLock
+        lock to guard against changes to `_layers_to_task` when finding or
+        removing tasks.
+
+    Methods
+    -------
+    slice_layers_async:
+        creates the slice and submits it to be executes
+    shutdown:
+        shutdown the executor on this class. Should only be called when the
+        class is no longer being used
+    _slice_layers:
+        convert request objects to slices
+    _on_slice_done:
+        callback executed when slicing future is complete
+    _try_to_remove_task:
+        utility function to remove a task from `_layers_to_task`
+    _find_existing_task:
+        utility function to find a task in `_layers_to_task`
```

Comment:
In general, I don't think we manually document methods in the class docstring, likely because it's a pain to keep these in sync. If you feel strongly about documenting them here, then I'd suggest only document the ones intended to be public (i.e. `slice_layers_async` and `shutdown`).